### DISCUSSION TITLE
Return value AX

### DISCUSSION THREAD

**Discussion by polluks** [cc65/cc65#1480] [General]
POV of the assembler: JOY_COUNT as 16 bit value, why not? I don't care.
POV of the compiler: I have a prototype `unsigned char __fastcall__ joy_count (void);` and 16 bit doesn't fit in char. So JOY_COUNT=1000 should not silently assemble. Not reasonable? We still have the same output.
https://github.com/cc65/cc65/issues/1478#issuecomment-828378940

**Comment by mrdudz** [comment]
You could always make a guard check for >JOY_COUNT not being 0. Kindof pointless though, given the probability of it every happening :) (And its still a constant, not a "16 bit value")

**Reply by polluks** [reply]
Well, the "guard check" is the code `lda #JOY_COUNT` itself.
foo.s(1): Error: Range error (1000 not in [0..255])
Right now we are suppressing this error, why?
I want to get rid of the pointless operators!

**Comment by mrdudz** [comment]
That is a good point :)

**Comment by polluks** [comment]
19 years ago Uz wrote "COUNT: Return the total number of available joysticks in a/x." but he meant "...in a and clear x."

**Comment by mrdudz** [comment]
I doubt it - i wrote that pce driver, it only clears X explicitly because there is an instruction for it in 6280 :)

**Reply by polluks** [reply]
I know, it was the first time code changes in this context.

**Comment by groessler** [comment]
> 19 years ago Uz wrote "COUNT: Return the total number of available joysticks in a/x." but he meant "...in a and clear x."

No. He meant return the number of joysticks in AX.

Think about having more than 255 joysticks connected (I know, it's currently a theoretical scenario),

**Reply by polluks** [reply]
Impossible because it's a char function, sorry. A char function in C always has to clear X because of side effects.

**Comment by mrdudz** [comment]
To be honest, i dont quite understand what the issue is here and what problem you are trying to solve :) Do you expect anyone to change JOY_COUNT to something bigger than 255 anytime soon? :)

**Reply by polluks** [reply]
Which expression is easier to understand `ldx #>JOY_COUNT` or `ldx #0`? Congratulations, you got the point.
Our groessler believed this function could return 16 bit, proven to be misleading.

**Comment by groessler** [comment]
> Impossible because it's a char function, sorry. A char function in C always has to clear X because of side effects.

I now see your point. Since it's a "char" function X should never be anything except 0. So, yes, "ldx #0" is in this context more appropriate.

**Comment by oliverschmidt** [comment]
@polluks. Okay, I now see your point and agree.

* If the function is prototyped as 16-bit value, then there should be `lda <???` and `ldx >???`.
* If the function is prototyped as 8-bit value, then there should be `lda ???` and `ldx #$00`.

To me it is very important to have an as consistent as possible code base. Therefore I'd like to see a PR that changes this in a consistent way.

**Reply by greg-king5** [reply]
>     * If the function is prototyped as 16-bit value, then there should be `lda <???` and `ldx >???`.
>     * If the function is prototyped as 8-bit value, then there should be `lda ???` and `ldx #$00`.
> 
> To me, it is very important to have an as consistent as possible code base. Therefore, I'd like to see a PR that changes this in a consistent way.

That's not consistent; that's two different ways of returning values.  To me, consistency means _always_ returning constants as:
```asm
lda #<???
ldx #>???
```
I say that this has little to do with a function's prototype.  Rather, it's an internal implementation detail.  The C standard says that expressions must calculate internally as INTs, not CHARs.  I think that Uz wanted to save space by having a function return 16 bits from one place, instead of having the compiler generate code to convert into 16 bits at every place that calls that function.

I don't want to see different coding styles because of different return types.  I want to see a consistent style because of a consistent implementation need.

EDIT:
In my mind,
```asm
lda #JOY_COUNT
ldx #$00
```
is misleading.  It looks as though the .X register isn't connected to the .A register.  It seems to be returning something else, maybe a flag, or perhaps even something as strange as an initializer for a loop counter that will be used by the caller.  I don't know, it isn't obvious -- it needs some comments.

On the other hand,
```asm
lda #<JOY_COUNT
ldx #>JOY_COUNT
```
is immediately obvious.  We can see right away that .X and .A work hand-in-hand.  We see that they form a _single_ container for a single object.  And, we see the endianess of that container.  Comments aren't needed.

To me, it's the opposite of misleading.

**Reply by spiro-trikaliotis** [reply]
Cite:
"The C standard says that expressions must calculate internally as INTs, not CHARs. I think that Uz wanted to save space by having a function return 16 bits from one place, instead of having the compiler generate code to convert into 16 bits at every place that calls that function."

To be more precise: In the standard, it is an "must behave as is".
That is, an implementation can decide to do it differently. It can ommit the promotion to char if the behaviour is still "as if the expressions wrote propageted to ints".

The easiest way to achieve this is to actually work with ints. If the compiler can do this better with chars, it is free to do so, as long as the result of the expression is the same. Of course, this is more work for the internals of the compiler.

**Comment by groessler** [comment]
I do understand Greg's comment, too. Which is different to what I wrote before.

Sigh. Life is complicated....

So I don't have a opinion about how it should be...

**Comment by mrdudz** [comment]
After thinking it over, i do agree with the original post. Its not about the standard or the function having to return X as 0 (that must be the case anyway). But - like polluks - i'd prefer an error when something is wrong over silently producing broken code. Right now when JOY_COUNT is > 255 it will result in broken code. if the function used lda #JOY_COUNT instead it would result in an error and make the problem immediatly obvious.

**Comment by groessler** [comment]
On 4/29/21 9:13 PM, spiro-trikaliotis wrote:
>
> Cite:
> "The C standard says that expressions must calculate internally as 
> INTs, not CHARs. I think that Uz wanted to save space by having a 
> function return 16 bits from one place, instead of having the compiler 
> generate code to convert into 16 bits at every place that calls that 
> function."
>

Can you give exact section number of the C standard? (I have 
ansi-c-99.pdf over here.)

regards,
chris

**Reply by spiro-trikaliotis** [reply]
Have a look at 5.1.2.3 "program execution".

"1 The  semantic  descriptions  in  this  International  Standard  describe  the  behavior  of  an abstract machine in which issues of optimization are irrelevant."

In "10 EXAMPLE 2", the there is even an actual example w.r.t. char to int promotions:

"In executing the fragment

char c1, c2;
/*...*/
c1 = c1 + c2;

the ‘‘integer promotions’’ require that the abstract machine promote the value of each variable to int size and then add the two ints and truncate the sum.  Provided the addition of two chars can be done without overflow,  or with  overflow wrapping  silently  to  produce  the  correct  result,  the  actual  execution  need  only produce the same result, possibly omitting the promotions."

**Comment by oliverschmidt** [comment]
> After thinking it over, i do agree with the original post. Its not about the standard or the function having to return X as 0 (that must be the case anyway).

Full ACK.

> But - like polluks - i'd prefer an error when something is wrong over silently producing broken code. Right now when JOY_COUNT is > 255 it will result in broken code. if the function used lda #JOY_COUNT instead it would result in an error and make the problem immediatly obvious.

FullACK.

These are the very thoughts I had too - and which made me change my mind.
